MoreLinq/SwapRange.cs

```diff
@@ -0,0 +1,85 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2017 Leandro F. Vieira (leandromoh). All rights reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Take a range of the source sequence and put it at other index. 
+        /// </summary>
+        /// <typeparam name="T">Type of the source sequence</typeparam>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="index">The zero-based index where the range to be swapped begins.</param>
+        /// <param name="count">The count of items to swap.</param>
+        /// <param name="putAt">The index where the specified range will be placed.</param>
+        /// <returns>The sequence with the specified range swapped to the specified position.</returns>
+        /// <remarks>
+        /// This operator uses deferred execution and streams its results.
+        /// </remarks>
+        /// <example>
+        /// <code>
+        /// var result = Enumerable.Range(0, 6).SwapRange(3, 2, 0);
+        /// </code>
+        /// The <c>result</c> variable will contain <c>{ 3, 4, 0, 1, 2, 5 }</c>.
+        /// </example>
+
+        public static IEnumerable<T> SwapRange<T>(this IEnumerable<T> source, int index, int count, int putAt)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (index < 0) throw new ArgumentOutOfRangeException(nameof(index), "Index cannot be negative.");
+            if (count < 0) throw new ArgumentOutOfRangeException(nameof(count), "Count cannot be negative.");
+            if (putAt < 0) throw new ArgumentOutOfRangeException(nameof(putAt), "PutAt cannot be negative.");
+
+            if (putAt == index || count == 0)
+                return source;
+
+            if (putAt < index)
```

Comment:
> In general, I find early exit pattern to be highly readable: list all 'special' conditions with their special return value, and then have the meat of the process without handling those cases.

I agree, that's how I do it but when I'm thinking _imperatively_ and wish to avoid code snake dance of complex if-else tree causing side-effects. The meat here, however, is in the local iterator function so what's left is a return expression. I think it's good to consider each case separately. By the way, #304 also [uses the conditional expression](https://github.com/leandromoh/MoreLINQ/blob/f0168dfc828d143efb16f90880e1ccebfef028d0/MoreLinq/TakeLast.cs#L55-L57).

Can we strike a compromise? :smile: Either make the tree obvious via indentation:

```c#
return newIndex == oldIndex || count == 0
     ? source
     : newIndex < oldIndex
       ? _(newIndex, oldIndex - newIndex, count)
       : _(oldIndex, count, newIndex - oldIndex);
```

or use the condition expression just for the call to the local function:

```c#
if (newIndex == oldIndex || count == 0)
    return source;

return newIndex < oldIndex
     ? _(newIndex, oldIndex - newIndex, count)
     : _(oldIndex, count, newIndex - oldIndex);
```

This would also make it consistent with [what we have today in `SkipLast`](https://github.com/morelinq/MoreLINQ/blob/v2.5.0/MoreLinq/SkipLast.cs#L40-L46).

And if it's still distasteful then I'm happy to let go.